lib/new-statemachine.fnl

```diff
@@ -0,0 +1,212 @@
+        ; TODO: Better name? This is the map that contains old, new, effect, etc.
+        ; TODO: Handle a signal with no handler
+        transition (tx-fn context action extra)
+        ;; _ (log.wf "XXX received transition info:\n%s" (hs.inspect transition)) ;; DELETEME
+        next-state transition.current-state
+        _ (log.wf "XXX next state: %s" next-state) ;; DELETEME
+        new-context transition.context
+        ;; _ (log.wf "XXX new context: %s" (hs.inspect new-context)) ;; DELETEME
+        effect transition.effect]
+    ; If next-state is nil, error: Means the action is not expected in this state
+    (log.wf "XXX Signal current: :%s next: :%s action: :%s extra: %s effect: :%s" current-state next-state action extra effect) ;; DELETEME
+
+    (set-current-state fsm next-state)
+    (set-context fsm new-context)
+    ; Call all subscribers
+    (each [_ sub (pairs (atom.deref fsm.subscribers))]
+      (sub {:context new-context :prev-state current-state :next-state next-state :effect effect :extra extra}))))
+
+(fn subscribe
+  [fsm sub]
+  "Adds a subscriber to the provided fsm. Returns a function to unsubscribe"
+  ; Super naive: Returns a function that just removes the entry at the inserted
+  ; key, but doesn't allow the same function to subscribe more than once since
+  ; its keyed by the string of the function itself.
+  (let [sub-key (tostring sub)]
+    (log.wf "Adding subscriber %s" sub) ;; DELETEME
+    (atom.swap! fsm.subscribers (fn [subs sub]
+                                  (merge {sub-key sub} subs)) sub)
+    ; Return the unsub func
+    (fn []
+      (atom.swap! fsm.subscribers (fn [subs key] (tset subs key nil)) sub-key))))
+
+(fn effect-handler
+  [effect-map]
+  "
+  Takes a map of effect->function and returns a function that handles these
+  effects by calling the mapped-to function, and then calls that function's
+  return value (a cleanup function) and calls it on the next transition.
+
+  These functions must return their own cleanup function or nil.
+  "
+  ;; Create a one-time atom used to store the cleanup function
+  (let [cleanup-ref (atom.new nil)]
+    ;; Return a subscriber function
+    (fn [{: context : prev-state : next-state : effect : action : extra}]
+      (log.wf "Effect handler called") ;; DELETEME
+      ;; Whenever a transition occurs, call the cleanup function, if set
+      (call-when (atom.deref cleanup-ref))
+      ;; Get a new cleanup function or nil and update cleanup-ref atom
+      (atom.reset! cleanup-ref
+                   (call-when (. effect-map effect) context extra)))))
+
+; TODO: We could require the initial state be a key in the states map
+; TODO: If we preserve the initial context we can maybe fsm.reset, though that's
+; hard to do safely since it only restores state and context, not the state of
+; hammerspoon itself, e.g. keys bindings, that have been messed with with all
+; the signal handlers.
+(fn create-machine
```

Comment:
I'd recommend having the API take a single table with the config as it is easier to maintain should any options be added. The other concern is that the initial state and initial context are specified in different places in this interface. Conceptually, the state and context are very tightly linked where systems using this library should be able to deterministically rely on state correctly updating context.

In languages like ReScript state and context can be specified as single value variants.

```rescript
type menu = {
  items: array<item>
}

type State =
  | Closed({ app: string })
  | Open({ app: string, menu: menu, history: array<menu> })
```

Which enforces that relationship given patten matching based on state is required to get the context.

Unfortunately in fennel, we don't get as strong guarantees so communicating that relationship in the API feels important here.

```fennel
(statemachine.new
  {:initial :closed
   :context {:menu {}
                   :history []}}
   :states {...}})
```

Given that the current-state and context are updated at exactly the same time, it seems like 1 atom would be enough. This makes it a bit easier to use externally:

```fennel
(local fsm (statemachine.new ...))

{: fsm }

;; In another module, advice, etc...

(let [{ : current-state : context} (atom.deref fsm.state)]
  (match current-state
    :open  context.menu
    :closed nil))
```

This suggestion is inspired by the XState createMachine API https://xstate.js.org/docs/packages/xstate-fsm/#super-quick-start

```fsm
const toggleMachine = createMachine({
  id: 'toggle',
  initial: 'inactive',
  states: {
    inactive: { on: { TOGGLE: 'active' } },
    active: { on: { TOGGLE: 'inactive' } }
  }
});
```

Additionally, I feel `new` would be preferable to `create-machine` here as it seems like a common hammerspoon and lua convention. This could then be leveraged to add a couple of methods:

#### subscribe
Which adds a subscriber function to the subscribers

```fennel
(local unsubscribe (fsm.subscribe (fn [{: prev : next : action : data : effect}] ...)))
```

#### get-state
Shorthand for `(atom.deref fsm.state)`

```fennel
(fsm.get-state)
;; =>
{:current-state :open
 :context {:menu {...} :history []}}
```

### Summary

- Recommending `create-machine` -> `new`
- Recommending `new` takes a single table for params that describe the system and highlight the relationship (open to suggestions on shape here)
- Recommending representing both :current-state and :context in a single atom so it can be derefed as a single value
- Recommending adding a `subscribe` method to the machine
- Recommending adding a `get-state` (open to better name ideas) method to the machine

Reply:
create-machine is actually taken from the existing naming. Note that it's renamed to `new` in the export.

The rest sounds good. I kept `subscribe`, `signal`, etc. as non-methods for now (passed the fsm object), happy to change that later.

Reply:
> create-machine is actually taken from the existing naming. Note that it's renamed to `new` in the export.

Hehe totally forgot about that. As long as it's exported as `new` for api consumers then internal name is fine, sorry for the contradiction there.

> The rest sounds good. I kept `subscribe`, `signal`, etc. as non-methods for now (passed the fsm object), happy to change that later.

Works for me

Reply:
Worth clarifying though is I was recommending having the methods in addition to the non-method API. There's advantages to both there

Reply:
All this is handled.